File: src/game_rules/action_parser.py
# -*- coding: utf-8 -*-
"""
Action Parser - LLM-слой для интерпретации свободных действий игрока
Использует данные SRD для определения возможности действия и параметров проверки
"""

from typing import Dict, Any, Optional, List
import json
from ai_engine.local_llm_client import LocalLLMClient, LocalLLMConfig
from .srd_data import SRDDataLoader
from prompts import DMPrompts


class ActionParser:
    """Парсер действий игрока в структурированный формат с использованием SRD данных"""
    
    def __init__(
        self, 
        llm_client: Optional[LocalLLMClient] = None, 
        parser_model: str = "llama3.1:8b",
        srd_loader: Optional[SRDDataLoader] = None
    ):
        """
        Инициализация парсера
        
        Args:
            llm_client: Клиент LLM для парсинга. Если None, создается новый.
            parser_model: Модель для парсинга. 
            srd_loader: Загрузчик SRD данных. Если None, создается новый.
        """
        if llm_client is None:
            config = LocalLLMConfig(
                model_name=parser_model,
                temperature=0.0,  # Максимальная детерминированность для парсинга
                max_tokens=400    # Увеличено для более детального ответа
            )
            self.llm_client = LocalLLMClient(config)
        else:
            self.llm_client = llm_client
        
        # Загружаем SRD данные для использования в промпте
        self.srd_loader = srd_loader or SRDDataLoader()
        # Данные будут загружаться динамически при парсинге
        self.skills_data = {}
        self.actions_data = {}
        self.dc_table = self.srd_loader.get_difficulty_table() or {}
    
    def parse_action(self, action_text: str, game_context: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Парсит действие игрока в структурированный JSON с использованием SRD данных
        Использует двухэтапный процесс:
        1. Модель выбирает нужные эндпоинты из SRD API
        2. Загружаются данные из выбранных эндпоинтов и передаются модели для финального парсинга
        
        Args:
            action_text: Текст действия игрока (например, "Я хочу перепрыгнуть реку")
            game_context: Контекст игры (локация, окружение, персонаж и т.д.)
        """
        try:
            # Этап 1: Выбор нужных эндпоинтов
            required_endpoints = self._select_required_endpoints(action_text)
            
            # Этап 2: Загрузка данных из выбранных эндпоинтов
            srd_data = self.srd_loader.load_multiple_endpoints(required_endpoints, limit_per_endpoint=20)
            
            # Обновляем кэш навыков для валидации
            if "skills" in srd_data:
                self.skills_data = {item.get("name", "").lower().replace("-", "_"): item for item in srd_data["skills"] if item.get("name")}
            
            # Этап 3: Финальный парсинг с данными из SRD
            system_prompt = self._get_parser_system_prompt()
            user_prompt = DMPrompts.get_action_parser_final_prompt(action_text, srd_data, game_context)
            
            messages = [{"role": "user", "content": user_prompt}]
            
            response = self.llm_client.generate_response(messages, system_prompt)
            parsed = self._extract_json_from_response(response, action_text)
            
            # Валидируем и дополняем результат
            parsed = self._validate_and_enrich_result(parsed, action_text)
            
            return parsed
        except Exception as e:
            # Возвращаем информацию об ошибке
            return {
                "is_possible": False,
                "error": f"Ошибка при парсинге действия: {str(e)}",
                "error_type": "exception",
                "intent": "unknown",
                "ability": "strength",
                "estimated_dc": 15,
                "reason": f"Не удалось обработать действие: {str(e)}. Попробуйте описать действие по-другому.",
                "action_text": action_text
            }
    
    def _select_required_endpoints(self, action_text: str) -> List[str]:
        """
        Этап 1: Выбор нужных эндпоинтов из SRD API
        
        Args:
            action_text: Текст действия игрока
            
        Returns:
            Список названий эндпоинтов, которые нужны для парсинга
        """
        # Получаем список доступных эндпоинтов
        available_endpoints = self.srd_loader.get_available_endpoints()
        
        # Промпт для выбора эндпоинтов
        system_prompt = """Ты — эксперт по правилам D&D 5e и структуре SRD API.

Твоя задача — проанализировать действие игрока и определить, какие эндпоинты из SRD API нужны для правильной интерпретации этого действия.

Отвечай ТОЛЬКО валидным JSON:
{
    "required_endpoints": ["skills", "ability-scores"]  // Список названий эндпоинтов
}"""
        
        user_prompt = DMPrompts.get_endpoint_selection_prompt(action_text, available_endpoints)
        messages = [{"role": "user", "content": user_prompt}]
        
        try:
            response = self.llm_client.generate_response(messages, system_prompt)
            parsed = self._extract_json_from_response(response, action_text)
            
            if parsed.get("error"):
                # Fallback: используем базовые эндпоинты
                return ["skills", "ability-scores", "rule-sections"]
            
            required_endpoints = parsed.get("required_endpoints", [])
            
            # Валидируем эндпоинты - проверяем, что они существуют
            valid_endpoints = []
            for endpoint in required_endpoints:
                if endpoint in available_endpoints.values() or endpoint in available_endpoints.keys():
                    valid_endpoints.append(endpoint)
            
            # Если нет валидных эндпоинтов, используем базовые
            if not valid_endpoints:
                return ["skills", "ability-scores", "rule-sections"]
            
            return valid_endpoints
        except Exception as e:
            print(f"⚠️ Ошибка при выборе эндпоинтов: {e}")
            # Fallback: используем базовые эндпоинты
            return ["skills", "ability-scores", "rule-sections"]

    def _get_parser_system_prompt(self) -> str:
        """Системный промпт для парсера действий с данными SRD"""
        # Формируем таблицу сложности (всегда нужна)
        if not self.dc_table:
            self.dc_table = self.srd_loader.get_difficulty_table() or {}
        dc_info = "\n".join([f"- {diff}: DC {dc}" for diff, dc in self.dc_table.items()]) if self.dc_table else ""
        
        # Используем промпт из DMPrompts (навыки будут в данных из эндпоинтов)
        return DMPrompts.get_action_parser_system_prompt(skills_list="", dc_info=dc_info)
    
    def _validate_and_enrich_result(self, parsed: Dict[str, Any], action_text: str) -> Dict[str, Any]:
        """
        Валидировать и обогатить результат парсинга
        
        Args:
            parsed: Результат парсинга от LLM
            action_text: Исходный текст действия
            
        Returns:
            Валидированный и обогащенный результат
        """
        # Если есть ошибка, возвращаем как есть
        if parsed.get("error"):
            return parsed
        
        # Убеждаемся, что все обязательные поля присутствуют
        result = {
            "is_possible": parsed.get("is_possible", True),
            "intent": parsed.get("intent", "unknown"),
            "ability": parsed.get("ability", "strength"),
            "skill": parsed.get("skill"),
            "estimated_dc": parsed.get("estimated_dc", "medium"),
            "estimated_difficulty": parsed.get("estimated_difficulty", "medium"),
            "modifiers": parsed.get("modifiers", []),
            "required_items": parsed.get("required_items", []),
            "reason": parsed.get("reason", ""),
            "base_action": parsed.get("base_action"),
            "action_text": action_text
        }
        
        # Валидируем навык - должен быть из SRD
        if result["skill"]:
            skill_normalized = result["skill"].lower().replace("-", "_")
            # Проверяем в кэше навыков (может быть словарь или список)
            skill_found = False
            if isinstance(self.skills_data, dict):
                if skill_normalized in self.skills_data:
                    skill_found = True
                else:
                    # Пробуем найти похожий навык
                    for skill_key in self.skills_data.keys():
                        if skill_key in skill_normalized or skill_normalized in skill_key:
                            result["skill"] = skill_key
                            skill_found = True
                            break
            elif isinstance(self.skills_data, list):
                # Если это список, ищем по name или index
                for skill_item in self.skills_data:
                    if isinstance(skill_item, dict):
                        skill_name = skill_item.get("name", "").lower().replace("-", "_")
                        skill_index = skill_item.get("index", "").lower().replace("-", "_")
                        if skill_normalized in skill_name or skill_normalized in skill_index:
                            result["skill"] = skill_index or skill_name
                            skill_found = True
                            break
            
            if not skill_found:
                # Навык не найден, но оставляем как есть (может быть null)
                result["skill"] = None
        
        # Валидируем DC - конвертируем строку в число если нужно
        if isinstance(result["estimated_dc"], str):
            if result["estimated_dc"] in self.dc_table:
                result["estimated_dc"] = self.dc_table[result["estimated_dc"]]
            else:
                # Используем среднюю сложность
                result["estimated_dc"] = self.dc_table.get("medium", 15)
        
        # Если действие невозможно, но нет reason, добавляем
        if not result["is_possible"] and not result["reason"]:
            result["reason"] = "Действие невозможно по правилам D&D 5e или нарушает законы физики/магии. Попробуйте описать другое действие."
        
        return result
    
    def _extract_json_from_response(self, response: str, action_text: str = "") -> Dict[str, Any]:
        """
        Извлечь JSON из ответа LLM
        
        Args:
            response: Ответ от LLM
            action_text: Исходный текст действия (для сообщения об ошибке)
            
        Returns:
            Словарь с распарсенными данными или словарь с ошибкой
        """
        if not response:
            return {
                "error": "Получен пустой ответ от LLM",
                "error_type": "empty_response"
            }
        
        response = response.strip()
        
        # Если ответ начинается с {, пытаемся распарсить
        if response.startswith('{'):
            try:
                parsed = json.loads(response)
                return parsed
            except json.JSONDecodeError as e:
                return {
                    "error": f"Ошибка парсинга JSON: {str(e)}",
                    "error_type": "json_decode_error"
                }
        
        # Ищем JSON в тексте
        start_idx = response.find('{')
        end_idx = response.rfind('}')
        
        if start_idx != -1 and end_idx != -1 and end_idx > start_idx:
            json_str = response[start_idx:end_idx + 1]
            try:
                parsed = json.loads(json_str)
                return parsed
            except json.JSONDecodeError as e:
                return {
                    "error": f"Ошибка парсинга JSON из текста: {str(e)}",
                    "error_type": "json_decode_error"
                }
        
        # JSON не найден в ответе
        return {
            "error": "Не удалось найти JSON в ответе LLM",
            "error_type": "json_not_found"
        }

File: src/main/java/com/dnd/game_rules/ActionParser.java
package com.dnd.game_rules;

import com.dnd.ai_engine.LocalLLMClient;
import com.dnd.prompts.DMPrompts;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import java.util.*;

/**
 * Парсер действий игрока в структурированный формат с использованием SRD данных
 * Использует двухэтапный процесс:
 * 1. Модель выбирает нужные эндпоинты из SRD API
 * 2. Загружаются данные из выбранных эндпоинтов и передаются модели для финального парсинга
 */
public class ActionParser {
    private static final Gson gson = new GsonBuilder().setLenient().create();
    private final LocalLLMClient llmClient;
    private final SRDDataLoader srdLoader;
    private Map<String, Object> skillsData = new HashMap<>();
    private Map<String, Integer> dcTable = new HashMap<>();

    public ActionParser(LocalLLMClient llmClient, SRDDataLoader srdLoader) {
        this.llmClient = llmClient;
        this.srdLoader = srdLoader;
        this.dcTable = srdLoader.getDifficultyTable();
    }

    public Map<String, Object> parseAction(String actionText, Map<String, Object> gameContext) {
        // Этап 1: Проверяем, требует ли действие проверки
        Map<String, Object> endpointSelection = selectRequiredEndpoints(actionText);
        
        Boolean requiresCheck = (Boolean) endpointSelection.get("requires_check");
        if (requiresCheck == null) requiresCheck = true;
        
        // Если действие не требует проверки, возвращаем результат без парсинга через SRD
        if (!requiresCheck) {
            System.out.println("✅ [ActionParser] Действие не требует проверки, пропускаем парсинг через SRD");
            Map<String, Object> result = new HashMap<>();
            result.put("is_possible", true);
            result.put("requires_dice_roll", false);
            result.put("intent", "trivial");
            result.put("ability", null);
            result.put("skill", null);
            result.put("estimated_dc", null);
            result.put("estimated_difficulty", null);
            result.put("modifiers", new ArrayList<>());
            result.put("required_items", new ArrayList<>());
            result.put("reason", "Тривиальное действие, не требует проверки навыка или характеристики");
            return result;
        }
        
        // Этап 2: Загрузка данных из выбранных эндпоинтов
        @SuppressWarnings("unchecked")
        List<String> requiredEndpoints = (List<String>) endpointSelection.get("required_endpoints");
        if (requiredEndpoints == null || requiredEndpoints.isEmpty()) {
            throw new RuntimeException("Действие требует проверки, но не указаны эндпоинты");
        }
        
        Map<String, List<Map<String, Object>>> srdData = srdLoader.loadMultipleEndpoints(requiredEndpoints);
        
        // Обновляем кэш навыков для валидации
        if (srdData.containsKey("skills")) {
            List<Map<String, Object>> skills = srdData.get("skills");
            skillsData.clear();
            for (Map<String, Object> skill : skills) {
                String name = (String) skill.getOrDefault("name", "");
                if (!name.isEmpty()) {
                    String normalized = name.toLowerCase().replace("-", "_");
                    skillsData.put(normalized, skill);
                }
            }
        }
        
        // Этап 3: Финальный парсинг с данными из SRD
        String systemPrompt = getParserSystemPrompt();
        String userPrompt = DMPrompts.getActionParserFinalPrompt(actionText, convertSRDData(srdData), gameContext);
        
        List<Map<String, String>> messages = new ArrayList<>();
        messages.add(Map.of("role", "user", "content", userPrompt));
        
        System.out.println("🤖 [ActionParser] Финальный парсинг действия: " + actionText);
        System.out.println("📤 [ActionParser] User prompt (первые 400 символов): " + userPrompt);
        
        String response = llmClient.generateResponse(messages, systemPrompt);
        
        System.out.println("📥 [ActionParser] Полный ответ LLM для парсинга действия:");
        System.out.println("   " + response);
        
        Map<String, Object> parsed = extractJsonFromResponse(response, actionText);
        
        // Валидируем и дополняем результат
        parsed = validateAndEnrichResult(parsed, actionText);
        
        return parsed;
    }
    
    private Map<String, Object> convertSRDData(Map<String, List<Map<String, Object>>> srdData) {
        Map<String, Object> result = new HashMap<>();
        for (Map.Entry<String, List<Map<String, Object>>> entry : srdData.entrySet()) {
            result.put(entry.getKey(), entry.getValue());
        }
        return result;
    }

    private Map<String, Object> selectRequiredEndpoints(String actionText) {
        // Получаем список доступных эндпоинтов
        Map<String, String> availableEndpoints = srdLoader.getAvailableEndpoints();
        
        // Используем промпты из ActionPrompts
        String systemPrompt = DMPrompts.getEndpointSelectionSystemPrompt();
        String userPrompt = DMPrompts.getEndpointSelectionPrompt(actionText, availableEndpoints);
        List<Map<String, String>> messages = new ArrayList<>();
        messages.add(Map.of("role", "user", "content", userPrompt));
        
        System.out.println("🤖 [ActionParser] Запрос выбора эндпоинтов для действия: " + actionText);
        
        String response = llmClient.generateResponse(messages, systemPrompt);
        
        System.out.println("📥 [ActionParser] Полный ответ LLM для выбора эндпоинтов:");
        System.out.println("   " + response);
        
        Map<String, Object> parsed = extractJsonFromResponse(response, actionText);
        
        if (parsed.containsKey("error")) {
            throw new RuntimeException("Ошибка при выборе эндпоинтов: " + parsed.get("error"));
        }
        
        // Проверяем requires_check
        Boolean requiresCheck = null;
        if (parsed.containsKey("requires_check")) {
            Object reqCheckObj = parsed.get("requires_check");
            System.out.println("🔍 [ActionParser] requires_check найден в ответе: " + reqCheckObj + " (тип: " + (reqCheckObj != null ? reqCheckObj.getClass().getSimpleName() : "null") + ")");
            if (reqCheckObj instanceof Boolean) {
                requiresCheck = (Boolean) reqCheckObj;
            } else if (reqCheckObj instanceof String) {
                requiresCheck = Boolean.parseBoolean((String) reqCheckObj);
            } else {
                System.err.println("⚠️ [ActionParser] Неожиданный тип для requires_check: " + reqCheckObj.getClass());
            }
        } else {
            System.out.println("⚠️ [ActionParser] requires_check отсутствует в ответе LLM");
        }
        
        if (requiresCheck == null) {
            // По умолчанию считаем, что требуется проверка
            System.out.println("⚠️ [ActionParser] requires_check = null, устанавливаем по умолчанию: true");
            requiresCheck = true;
        } else {
            System.out.println("✅ [ActionParser] requires_check = " + requiresCheck);
        }
        
        parsed.put("requires_check", requiresCheck);
        
        // Валидируем эндпоинты только если требуется проверка
        if (requiresCheck) {
            @SuppressWarnings("unchecked")
            List<String> requiredEndpoints = (List<String>) parsed.get("required_endpoints");
            
            if (requiredEndpoints == null || requiredEndpoints.isEmpty()) {
                throw new RuntimeException("Действие требует проверки, но LLM не вернул список требуемых эндпоинтов. Ответ: " + response);
            }
            
            // Валидируем эндпоинты - проверяем, что они существуют
            List<String> validEndpoints = new ArrayList<>();
            for (String endpoint : requiredEndpoints) {
                if (availableEndpoints.containsKey(endpoint) || availableEndpoints.containsValue(endpoint)) {
                    validEndpoints.add(endpoint);
                }
            }
            
            if (validEndpoints.isEmpty()) {
                throw new RuntimeException("Не найдено ни одного валидного эндпоинта из запрошенных: " + requiredEndpoints);
            }
            
            parsed.put("required_endpoints", validEndpoints);
        } else {
            // Если не требуется проверка, эндпоинты не нужны
            parsed.put("required_endpoints", new ArrayList<>());
        }
        
        return parsed;
    }
    
    private String getParserSystemPrompt() {
        // Формируем таблицу сложности (всегда нужна)
        if (dcTable.isEmpty()) {
            dcTable = srdLoader.getDifficultyTable();
        }
        StringBuilder dcInfo = new StringBuilder();
        for (Map.Entry<String, Integer> entry : dcTable.entrySet()) {
            dcInfo.append(String.format("- %s: DC %d\n", entry.getKey(), entry.getValue()));
        }
        
        // Используем промпт из DMPrompts (навыки будут в данных из эндпоинтов)
        return DMPrompts.getActionParserSystemPrompt("", dcInfo.toString());
    }

    private Map<String, Object> validateAndEnrichResult(Map<String, Object> parsed, String actionText) {
        // Если есть ошибка, выбрасываем исключение
        if (parsed.containsKey("error")) {
            throw new RuntimeException("Ошибка в результате парсинга: " + parsed.get("error"));
        }
        
        // Убеждаемся, что все обязательные поля присутствуют
        Map<String, Object> result = new HashMap<>();
        result.put("is_possible", parsed.getOrDefault("is_possible", true));
        // requires_dice_roll: если не указан, по умолчанию нужен бросок
        // Но если указан явно (даже как false), используем его значение
        if (parsed.containsKey("requires_dice_roll")) {
            result.put("requires_dice_roll", parsed.get("requires_dice_roll"));
        } else {
            result.put("requires_dice_roll", true); // По умолчанию нужен бросок
        }
        result.put("intent", parsed.getOrDefault("intent", "unknown"));
        result.put("ability", parsed.getOrDefault("ability", "strength"));
        result.put("skill", parsed.get("skill"));
        result.put("estimated_dc", parsed.getOrDefault("estimated_dc", "medium"));
        result.put("estimated_difficulty", parsed.getOrDefault("estimated_difficulty", "medium"));
        result.put("modifiers", parsed.getOrDefault("modifiers", new ArrayList<>()));
        result.put("required_items", parsed.getOrDefault("required_items", new ArrayList<>()));
        result.put("reason", parsed.getOrDefault("reason", ""));
        result.put("action_text", actionText);
        
        // Валидируем навык - должен быть из SRD
        if (result.get("skill") != null) {
            String skill = (String) result.get("skill");
            String skillNormalized = skill.toLowerCase().replace("-", "_");
            
            // Проверяем в кэше навыков
            boolean skillFound = false;
            if (skillsData.containsKey(skillNormalized)) {
                skillFound = true;
            } else {
                // Пробуем найти похожий навык
                for (String skillKey : skillsData.keySet()) {
                    if (skillKey.contains(skillNormalized) || skillNormalized.contains(skillKey)) {
                        result.put("skill", skillKey);
                        skillFound = true;
                        break;
                    }
                }
            }
            
            if (!skillFound) {
                // Навык не найден, но оставляем как есть (может быть null)
                result.put("skill", null);
            }
        }
        
        // Валидируем DC - конвертируем строку в число если нужно
        Object estimatedDc = result.get("estimated_dc");
        if (estimatedDc instanceof String) {
            String dcStr = (String) estimatedDc;
            if (dcTable.containsKey(dcStr)) {
                result.put("estimated_dc", dcTable.get(dcStr));
            } else {
                // Используем среднюю сложность
                result.put("estimated_dc", dcTable.getOrDefault("medium", 15));
            }
        }
        
        // Если действие невозможно, но нет reason, добавляем
        if (!(Boolean) result.get("is_possible") && ((String) result.get("reason")).isEmpty()) {
            result.put("reason", "Действие невозможно по правилам D&D 5e или нарушает законы физики/магии. Попробуйте описать другое действие.");
        }
        
        return result;
    }
    
    private Map<String, Object> extractJsonFromResponse(String response, String actionText) {
        if (response == null || response.trim().isEmpty()) {
            System.err.println("❌ [ActionParser] Получен пустой ответ от LLM для действия: " + actionText);
            throw new RuntimeException("Получен пустой ответ от LLM при парсинге действия: " + actionText);
        }
        
        response = response.trim();
        
        // Если ответ начинается с {, пытаемся распарсить
        if (response.startsWith("{")) {
            try {
                JsonObject jsonObj = gson.fromJson(response, JsonObject.class);
                Map<String, Object> result = parseJsonObject(jsonObj);
                return result;
            } catch (Exception e) {
                System.err.println("❌ [ActionParser] Ошибка парсинга JSON: " + e.getMessage());
                System.err.println("   Полный ответ: " + response);
                throw new RuntimeException("Ошибка парсинга JSON из ответа LLM: " + e.getMessage() + ". Ответ: " + response, e);
            }
        }
        
        // Ищем JSON в тексте
        System.out.println("🔍 [ActionParser] Поиск JSON в тексте ответа");
        int startIdx = response.indexOf('{');
        int endIdx = response.lastIndexOf('}');
        
        if (startIdx == -1 || endIdx == -1 || endIdx <= startIdx) {
            System.err.println("❌ [ActionParser] Не удалось найти JSON в ответе");
            System.err.println("   Полный ответ: " + response);
            throw new RuntimeException("Не удалось найти JSON в ответе LLM. Ответ: " + response);
        }
        
        String jsonStr = response.substring(startIdx, endIdx + 1);
        System.out.println("🔍 [ActionParser] Извлеченный JSON: " + jsonStr);
        try {
            JsonObject jsonObj = gson.fromJson(jsonStr, JsonObject.class);
            Map<String, Object> result = parseJsonObject(jsonObj);
            System.out.println("✅ [ActionParser] JSON успешно распарсен из текста");
            return result;
        } catch (RuntimeException e) {
            throw e;
        } catch (Exception e) {
            System.err.println("❌ [ActionParser] Ошибка парсинга извлеченного JSON: " + e.getMessage());
            System.err.println("   Извлеченный JSON: " + jsonStr);
            throw new RuntimeException("Ошибка парсинга JSON из текста: " + e.getMessage() + ". JSON: " + jsonStr, e);
        }
    }
    
    private Map<String, Object> parseJsonObject(JsonObject jsonObj) {
        Map<String, Object> result = new HashMap<>();
        
        if (jsonObj.has("is_possible")) {
            result.put("is_possible", jsonObj.get("is_possible").getAsBoolean());
        }
        
        if (jsonObj.has("requires_check")) {
            if (jsonObj.get("requires_check").isJsonNull()) {
                result.put("requires_check", true); // По умолчанию требуется проверка
            } else {
                result.put("requires_check", jsonObj.get("requires_check").getAsBoolean());
            }
        }
        
        if (jsonObj.has("requires_dice_roll")) {
            if (jsonObj.get("requires_dice_roll").isJsonNull()) {
                result.put("requires_dice_roll", true); // По умолчанию нужен бросок
            } else {
                result.put("requires_dice_roll", jsonObj.get("requires_dice_roll").getAsBoolean());
            }
        }
        
        if (jsonObj.has("intent")) {
            if (jsonObj.get("intent").isJsonNull()) {
                result.put("intent", null);
            } else {
                result.put("intent", jsonObj.get("intent").getAsString());
            }
        }
        
        if (jsonObj.has("ability")) {
            if (jsonObj.get("ability").isJsonNull()) {
                result.put("ability", null);
        } else {
                result.put("ability", jsonObj.get("ability").getAsString());
            }
        }
        
        if (jsonObj.has("skill")) {
            if (jsonObj.get("skill").isJsonNull()) {
            result.put("skill", null);
            } else {
                result.put("skill", jsonObj.get("skill").getAsString());
            }
        }
        
        if (jsonObj.has("estimated_dc")) {
            if (jsonObj.get("estimated_dc").isJsonNull()) {
                result.put("estimated_dc", null);
            } else if (jsonObj.get("estimated_dc").isJsonPrimitive()) {
                if (jsonObj.get("estimated_dc").getAsJsonPrimitive().isNumber()) {
                    result.put("estimated_dc", jsonObj.get("estimated_dc").getAsInt());
                } else {
                    result.put("estimated_dc", jsonObj.get("estimated_dc").getAsString());
                }
            }
        }
        
        if (jsonObj.has("estimated_difficulty")) {
            if (jsonObj.get("estimated_difficulty").isJsonNull()) {
                result.put("estimated_difficulty", null);
            } else {
                result.put("estimated_difficulty", jsonObj.get("estimated_difficulty").getAsString());
            }
        }
        
        if (jsonObj.has("modifiers")) {
            if (jsonObj.get("modifiers").isJsonNull()) {
        result.put("modifiers", new ArrayList<>());
            } else {
                List<String> modifiers = new ArrayList<>();
                jsonObj.getAsJsonArray("modifiers").forEach(e -> {
                    if (!e.isJsonNull()) {
                        modifiers.add(e.getAsString());
                    }
                });
                result.put("modifiers", modifiers);
            }
        }
        
        if (jsonObj.has("required_items")) {
            if (jsonObj.get("required_items").isJsonNull()) {
        result.put("required_items", new ArrayList<>());
            } else {
                List<String> items = new ArrayList<>();
                jsonObj.getAsJsonArray("required_items").forEach(e -> {
                    if (!e.isJsonNull()) {
                        items.add(e.getAsString());
                    }
                });
                result.put("required_items", items);
            }
        }
        
        if (jsonObj.has("reason")) {
            if (jsonObj.get("reason").isJsonNull()) {
                result.put("reason", "");
            } else {
                result.put("reason", jsonObj.get("reason").getAsString());
            }
        }
        
        if (jsonObj.has("required_endpoints")) {
            if (jsonObj.get("required_endpoints").isJsonNull()) {
                result.put("required_endpoints", new ArrayList<>());
            } else {
                List<String> endpoints = new ArrayList<>();
                jsonObj.getAsJsonArray("required_endpoints").forEach(e -> {
                    if (!e.isJsonNull()) {
                        endpoints.add(e.getAsString());
                    }
                });
                result.put("required_endpoints", endpoints);
            }
        }
        
        return result;
    }


}

